### ISSUE TITLE
T3D: Train YOLOv8/YOLO11 & Export ONNX 🧠

### ISSUE THREAD

**Issue by vikraman2212** [vikraman2212/sports-analyst#24]
# Task 3D: Train YOLOv8/YOLO11 & Export ONNX

**Status:** 📋 PLANNED  
**Parent:** #4 (T3: YOLOv8 Migration)  
**Duration:** 3 days  
**Start Date:** 2025-11-05  
**Owner:** ML  
**Dependencies:** T3C (#23)

---

## 🎯 Objective

Train custom YOLOv8 or YOLO11 model on cricket ball dataset, optimize for mobile/web deployment, and export to ONNX format for browser inference.

## 📦 Deliverables

- [ ] **Best Model Checkpoint:** `best.pt` (PyTorch format)
- [ ] **ONNX Export:** `best.onnx` (opset ≥17 for onnxruntime-web)
- [ ] **Training Logs:** Metrics, loss curves, validation results
- [ ] **Export Verification:** ONNX model loads in `onnxruntime-web`
- [ ] **Documentation:** Training config, hyperparameters, export notes

## ✅ Acceptance Criteria

- [ ] **Recall ≥ 0.85** on holdout ball frames (high detection rate)
- [ ] **FPS ≥ 30 @ 720p** (or acceptable with frame sampling)
- [ ] ONNX model loads successfully in `onnxruntime-web`
- [ ] Model size < 10MB (for fast web loading)
- [ ] Training logs show convergence (no overfitting)

### Model Selection

**Option 1: YOLOv8n (Nano)** - Fastest
- Size: ~6MB
- Speed: ~50 FPS @ 720p (mobile GPU)
- Accuracy: Good for single-class

**Option 2: YOLOv8s (Small)** - Balanced
- Size: ~22MB
- Speed: ~35 FPS @ 720p
- Accuracy: Better for small objects

**Option 3: YOLO11n (Latest)** - Experimental
- Size: Similar to v8n
- Speed: ~10% faster than v8n
- Accuracy: Improved architecture

**Recommendation:** Start with **YOLOv8n**, upgrade to YOLOv8s if recall < 0.85.

### Hyperparameters

```yaml
# training_config.yaml
model: yolov8n.yaml  # or yolo11n.yaml
data: cricket_ball.yaml

epochs: 100
batch: 16  # Adjust based on GPU memory
imgsz: 640  # Input image size

# Optimizer
optimizer: AdamW
lr0: 0.001  # Initial learning rate
lrf: 0.01   # Final learning rate

# Augmentation (from T3C)
mosaic: 1.0
copy_paste: 0.3
scale: 0.5
hsv_h: 0.015
hsv_s: 0.7
hsv_v: 0.4

# Small object focus
anchor_t: 4.0  # Anchor threshold
iou: 0.7       # IoU threshold for NMS
```

### Training Script

```python
from ultralytics import YOLO

# Load base model
model = YOLO('yolov8n.yaml')

# Train
results = model.train(
    data='cricket_ball.yaml',
    epochs=100,
    imgsz=640,
    batch=16,
    device=0,  # GPU 0 (use 'cpu' if no GPU)
    name='cricket_ball_v1',
    patience=20,  # Early stopping
    save_period=10,  # Save checkpoint every 10 epochs
)

# Validate
metrics = model.val()
print(f"Recall: {metrics.box.r}")
print(f"Precision: {metrics.box.p}")
print(f"mAP50: {metrics.box.map50}")
```

### Export Script

```python
from ultralytics import YOLO

# Load best checkpoint
model = YOLO('runs/detect/cricket_ball_v1/weights/best.pt')

# Export to ONNX
model.export(
    format='onnx',
    opset=17,  # Required for onnxruntime-web
    simplify=True,  # Simplify graph
    dynamic=False,  # Fixed input size for web
    imgsz=640,
)

# Output: best.onnx
```

### Verification (Python)

```python
import onnxruntime as ort
import numpy as np

# Load ONNX model
session = ort.InferenceSession('best.onnx')

# Test inference
dummy_input = np.random.randn(1, 3, 640, 640).astype(np.float32)
outputs = session.run(None, {'images': dummy_input})

print("ONNX model loaded successfully!")
print(f"Output shapes: {[o.shape for o in outputs]}")
```

### Verification (JavaScript)

```javascript
import * as ort from 'onnxruntime-web';

const session = await ort.InferenceSession.create('best.onnx');
const input = new ort.Tensor('float32', new Float32Array(1*3*640*640), [1, 3, 640, 640]);
const outputs = await session.run({ images: input });

console.log('ONNX model loaded in browser!');
```

### Training Metrics
- **Loss:** Box loss, class loss, DFL loss
- **mAP50:** Mean average precision @ IoU 0.5
- **mAP50-95:** mAP across IoU thresholds
- **Precision:** TP / (TP + FP)
- **Recall:** TP / (TP + FN)

### Validation Metrics (Holdout Set)
- **Recall ≥ 0.85** (primary target)
- **Precision ≥ 0.80**
- **FPS @ 720p:** Measure on target device

### Overfitting Check
- Training loss vs. validation loss
- mAP50 plateau (should converge)

## 📁 Output Files

Create in `ml/training/`:
- `runs/detect/cricket_ball_v1/` - Training outputs
  - `weights/best.pt` - Best checkpoint
  - `weights/last.pt` - Final checkpoint
  - `results.csv` - Training metrics
  - `confusion_matrix.png`
  - `results.png` - Loss/mAP curves

Export to `frontend/public/models/`:
- `cricket_ball_v1.onnx` - Production model

Create in `docs/`:
- `T3D_training_report.md` - Full training results
- `T3D_export_notes.md` - ONNX export instructions

### Minimum
- **GPU:** NVIDIA GTX 1060 (6GB VRAM) or equivalent
- **RAM:** 16GB
- **Storage:** 10GB free space

### Recommended
- **GPU:** NVIDIA RTX 3060 (12GB VRAM)
- **RAM:** 32GB
- **Storage:** 50GB SSD

### Cloud Alternative
- **Google Colab:** Free T4 GPU (limit: 12 hours)
- **Kaggle:** Free P100 GPU (limit: 30 hours/week)

### Issue: Low Recall (< 0.85)
**Solution:**
1. Increase training epochs (100 → 150)
2. Upgrade to YOLOv8s (larger model)
3. Add more training data (augment dataset)
4. Tune confidence threshold (lower for recall)

### Issue: Slow FPS (< 30)
**Solution:**
1. Use YOLOv8n instead of YOLOv8s
2. Reduce input size (640 → 480)
3. Implement frame sampling (process every 2nd frame)

### Issue: ONNX Export Fails
**Solution:**
1. Check PyTorch version compatibility
2. Use `opset=17` (not higher)
3. Set `dynamic=False` for web

## 🔗 Dependencies

**Upstream:** T3C (Dataset Plan) - dataset must be ready  
**Downstream:** T3E (Integration) - use exported ONNX model

## 📚 References

- YOLOv8 Training: https://docs.ultralytics.com/modes/train/
- YOLO11 Release: https://docs.ultralytics.com/models/yolo11/
- ONNX Export: https://docs.ultralytics.com/integrations/onnx/
- onnxruntime-web: https://onnxruntime.ai/docs/tutorials/web/

---

**Epic:** #4 (T3)  
**Previous:** T3C (Dataset Plan)  
**Next:** T3E (Integration)  
**Estimated Training Time:** 3-6 hours (GPU dependent)
